Dip/Dip/Definition.swift

```diff
@@ -50,20 +50,62 @@ func ==(lhs: DefinitionKey, rhs: DefinitionKey) -> Bool {
 public enum ComponentScope {
   /// Indicates that a new instance of the component will be created each time it's resolved.
   case Prototype
+  /// Indicates that instances will be reused during resolve but will be discurded when topmost `resolve` method returns.
+  case ObjectGraph
   /// Indicates that resolved component should be retained by container and always reused.
   case Singleton
 }
 
 ///Definition of type T describes how instances of this type should be created when they are resolved by container.
 public final class DefinitionOf<T>: Definition {
```

Comment:
I was thinking (didn't think it through, that's just a though), couldn't `DefinitionOf`be a `struct` to leverage value type?
Then `inScope` would return a new `DefinitionOf` instead of changing it in place (which seems less FP)?